presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSource.java

```diff
@@ -92,11 +103,35 @@ public boolean isFinished()
     @Override
     public Page getNextPage()
     {
+        boolean supplyRowIDs = recordReader.isColumnPresent(ROW_ID_COLUMN_INDEX);
         try {
             Page page = recordReader.getNextPage();
             if (page == null) {
                 close();
             }
+            else if (supplyRowIDs) { // If we need a row ID block, synthesize it here.
+                // rowNumbers is always the last block in the page
+                Block rowNumbers = page.getBlock(page.getChannelCount() - 1);
+                Block rowIDs = coercer.apply(rowNumbers);
+
+                // figure out which block is the row ID and replace it
+                int rowIDColumnIndex = recordReader.toZeroBasedColumnIndex(ROW_ID_COLUMN_INDEX);
```

Comment:
This needs to happen in the constructor.

I was under the impression that we do have the rowIDColumnIndex in the page source factory, and we can just pass it into the page source?